lib/engine/chelsea_bridge.ex

```diff
@@ -2,56 +2,127 @@ defmodule Engine.ChelseaBridge do
+    now = Timex.now()
+
+    token =
+      if(state.token.expiration == nil or DateTime.compare(state.token.expiration, now) == :lt) do
+        update_api_token(now)
+      else
+        state.token
+      end
 
     with {:ok, %{status_code: 200, body: body}} <-
-           http_client.get(url, [{"Authorization", "Basic #{auth}"}]),
-         {:ok, data} <- Jason.decode(body),
-         %{"bridge" => %{"bridgeStatusId" => %{"status" => status}}} <- data do
-      estimate =
-        case data do
-          %{"lift_estimate" => %{"estimate_time" => estimate}} ->
-            String.replace(estimate, ~r/\.\d+$/, "")
-            |> Timex.parse!("{YYYY}-{M}-{D} {h24}:{m}:{s}")
-            |> DateTime.from_naive!("America/New_York")
-
-          _ ->
-            nil
-        end
-
-      :ets.insert(:bridge_status, {:value, %{raised?: status == "Raised", estimate: estimate}})
+           http_client.get("#{@base_api_url}#{@api_status_endpoint}", [
+             {"Authorization", "Bearer #{token.value}"}
+           ]),
+         {:ok, data} <- Jason.decode(body) do
+      EtsUtils.write_ets(
+        state.table,
+        %{
+          :value => %{
+            raised?: Map.get(data, "liftInProgress"),
+            estimate: DateTime.add(now, Map.get(data, "estimatedDurationInMinutes"), :minute)
+          }
+        },
+        :none
+      )
+
+      {:noreply, %{state | token: token}}
     else
       err ->
         Logger.error("Error getting bridge status: #{inspect(err)}")
     end
-
-    {:noreply, state}
   end
 
   def handle_info(msg, state) do
     Logger.warning("Engine.ChelseaBridge unknown_message: #{inspect(msg)}")
     {:noreply, state}
   end
+
+  @spec update_api_token(DateTime.t()) :: token
+  def update_api_token(now) do
+    # Calls a token granting endpoint with a username and password from our environment variables.
+    # This token grants Bridge Data API access, and must be refreshed at least every month
+    username = Application.get_env(:realtime_signs, :chelsea_bridge_username)
+    password = Application.get_env(:realtime_signs, :chelsea_bridge_password)
+    http_poster = Application.get_env(:realtime_signs, :http_poster_mod)
+
+    # Encode as application/x-www-form-urlencoded
+    body =
+      URI.encode_query(%{
+        "grant_type" => "password",
+        "username" => username,
+        "password" => password
+      })
+
+    with {:ok, %{status_code: 200, body: body}} <-
+           http_poster.post("#{@base_api_url}#{@api_token_endpoint}", body, [
+             {"Content-Type", "application/x-www-form-urlencoded"}
+           ]),
+         {:ok, data} <- Jason.decode(body) do
+      expiration =
+        now
+        |> DateTime.add(Map.get(data, "expires_in"), :second)
+
+      %{
+        value: Map.get(data, "access_token"),
+        expiration: expiration
+      }
+    else
+      err ->
+        Logger.error("Error getting bridge access_token: #{inspect(err)}")
```

Comment:
We should return `%{value: nil, expiration: nil}` here, to avoid crashing when we try to use the token. It'll still try to make the call with a blank token, and the server should send a failure response, but that seems fine.